src/dependency-manager/src/utils/interpolation/parameter-interpolator.ts

```diff
@@ -0,0 +1,119 @@
+import Mustache from 'mustache';
+import { ServiceNode } from '../..';
+import DependencyGraph from '../../graph';
+import { InterpolationContext } from '../../interpolation/interpolation-context';
+import { ParameterValueV2 } from '../../service-config/base';
+import { EnvironmentInterfaceContext, EnvironmentInterpolationContext, ServiceInterfaceContext } from './interpolation-context';
+
+export class ParameterInterpolator {
+
+  public static interpolateString(param_value: string, environment_context: EnvironmentInterpolationContext): string {
+    Mustache.tags = ['${', '}']; // sets custom delimiters
+    Mustache.escape = function (text) { return text; } // turns off HTML escaping
+    //TODO:77: add validation logic to catch expressions that don't refer to an existing path
+    return Mustache.render(param_value, environment_context);
+  }
+
+  public static build_friendly_name_map(graph: DependencyGraph): { [key: string]: { [key: string]: string } } {
+    const friendly_name_map: { [key: string]: { [key: string]: string } } = {};
+    for (const node of graph.nodes) {
+      if (node instanceof ServiceNode) {
+        friendly_name_map[node.ref] = {};
+        for (const friendly_name of Object.keys(node.node_config.getDependencies())) {
+          const top_level_node = graph.nodes
+            .filter(n => n instanceof ServiceNode)
+            .map(n => n as ServiceNode)
+            .find(n => n.node_config.getName() === friendly_name);
+
+          if (top_level_node) {
+            friendly_name_map[node.ref][friendly_name] = top_level_node.namespace_ref;
+          }
+        }
+      }
+    }
+    return friendly_name_map;
+  }
+
+  public static mapGraphToInterpolationContext(graph: DependencyGraph, interface_context: EnvironmentInterfaceContext): EnvironmentInterpolationContext {
+    const environment_context: EnvironmentInterpolationContext = {};
+
+    for (const node of graph.nodes.filter(n => n instanceof ServiceNode).map(n => n as ServiceNode)) {
+      const service_context = ParameterInterpolator.mapNodeToInterpolationContext(node, interface_context[node.ref]);
+      environment_context[node.namespace_ref] = service_context;
+    }
+
+    return environment_context;
+  }
+
+  public static mapNodeToInterpolationContext(node: ServiceNode, interface_context: ServiceInterfaceContext): InterpolationContext {
+    return {
+      parameters: Object.entries(node.node_config.getParameters())
```

Comment:
I think I see why you defined InterpolationContext separately. I assume converting the config to json isn't enough since we transform in the getters?

Reply:
This is just one example that underlies the real problem mentioned above: ServiceConfig represents the shape of the architect.json that the user is allowed to give us. I'm not sure we want to overload its use case... that will inevitably lead us to add cruft to the the ServiceConfig which will only be used in one context or another and will inevitably lead to conflicting requirements or a convoluted object that tries to serve two purposes poorly instead of one well.